File: .API/Models/UserPatreonData.cs
// Decompiled with JetBrains decompiler
// Type: CloudX.Shared.UserPatreonData
// Assembly: CloudX.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 6223B97A-06A5-46CB-9E10-78604961D6EE
// Assembly location: J:\D\SteamLibrary\steamapps\common\NeosVR\HeadlessClient\CloudX.Shared.dll

using Newtonsoft.Json;
using System;
using System.Text.Json.Serialization;

namespace CloudX.Shared
{
  [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
  public class UserPatreonData
  {
    public const int MIN_WORLD_ACCESS_CENTS = 600;
    public const int ACTIVATION_LENGTH = 40;

    [NullOnExternal]
    [JsonProperty(PropertyName = "email")]
    [JsonPropertyName("email")]
    public string Email { get; set; }

    [JsonProperty(PropertyName = "isPatreonSupporter")]
    [JsonPropertyName("isPatreonSupporter")]
    public bool IsPatreonSupporter { get; set; }

    [JsonProperty(PropertyName = "lastPatreonPledgeCents")]
    [JsonPropertyName("lastPatreonPledgeCents")]
    public int LastPatreonPledgeCents { get; set; }

    [JsonProperty(PropertyName = "lastTotalCents")]
    [JsonPropertyName("lastTotalCents")]
    public int LastTotalCents { get; set; }

    [JsonProperty(PropertyName = "rewardMultiplier")]
    [JsonPropertyName("rewardMultiplier")]
    public double? RewardMultiplier { get; set; }

    [JsonProperty(PropertyName = "rewardType")]
    [JsonPropertyName("rewardType")]
    public string RewardType { get; set; }

    [JsonProperty(PropertyName = "customTier")]
    [JsonPropertyName("customTier")]
    public string CustomTier { get; set; }

    [Obsolete]
    [JsonProperty(PropertyName = "lastPlusActivationTime")]
    [JsonPropertyName("lastPlusActivationTime")]
    public DateTime LastPlusActivationTime
    {
      get
      {
        return this.LastActivationTime;
      }
      set
      {
        this.LastActivationTime = value;
      }
    }

    [JsonProperty(PropertyName = "lastActivationTime")]
    [JsonPropertyName("lastActivationTime")]
    public DateTime LastActivationTime { get; set; }

    [Obsolete]
    [JsonProperty(PropertyName = "lastPlusPledgeAmount")]
    [JsonPropertyName("lastPlusPledgeAmount")]
    public int LastPlusPledgeAmount
    {
      get
      {
        return this.LastPaidPledgeAmount;
      }
      set
      {
        this.LastPaidPledgeAmount = value;
      }
    }

    [JsonProperty(PropertyName = "lastPaidPledgeAmount")]
    [JsonPropertyName("lastPaidPledgeAmount")]
    public int LastPaidPledgeAmount { get; set; }

    public string AccountName
    {
      get
      {
        if (this.CustomTier != null)
          return this.CustomTier;
        return NeosAccount.AccountName(this.CurrentAccountType);
      }
    }

    public AccountType CurrentAccountType
    {
      get
      {
        if ((DateTime.UtcNow - this.LastActivationTime).TotalDays <= 40.0)
          return UserPatreonData.GetAccountType(this.LastPaidPledgeAmount);
        return AccountType.Normal;
      }
    }

    public AccountType PledgedAccountType
    {
      get
      {
        return UserPatreonData.GetAccountType(this.LastPatreonPledgeCents);
      }
    }

    public bool UpdatePatreonStatus(int currentTotalCents, out bool extendedPlus)
    {
      extendedPlus = false;
      int num = currentTotalCents - this.LastTotalCents;
      if (num <= 0)
      {
        if (this.LastActivationTime.Year > 2016)
          return false;
        num = this.LastPaidPledgeAmount;
      }
      if (num > 0)
      {
        this.LastActivationTime = DateTime.UtcNow;
        this.LastPaidPledgeAmount = num;
        extendedPlus = true;
      }
      this.LastTotalCents = currentTotalCents;
      return true;
    }

    private static AccountType GetAccountType(int cents)
    {
      for (AccountType type = AccountType.Anorak; type >= AccountType.Normal; --type)
      {
        if (cents >= NeosAccount.MinCents(type))
          return type;
      }
      return AccountType.Normal;
    }

    [JsonIgnore]
    [JsonIgnore]
    public bool HasPledgedEnoughForPlus
    {
      get
      {
        return Math.Max(this.LastPatreonPledgeCents, this.LastPaidPledgeAmount) >= NeosAccount.MinCents(AccountType.BladeRunner);
      }
    }

    [JsonIgnore]
    [JsonIgnore]
    public bool HasPledgedEnoughForWorlds
    {
      get
      {
        return Math.Max(this.LastPatreonPledgeCents, this.LastPaidPledgeAmount) >= 600;
      }
    }
  }
}

File: API/UserPatreonData.js
const { AccountType } = require("./AccountType");
const { List } = require("./List");
const { NeosAccount } = require("./NeosAccount");
const { PatreonSnapshot } = require("./PatreonSnapshot");
class UserPatreonData {
	constructor($b) {
		if (!$b) $b = {};
		this.MIN_WORLD_ACCESS_CENTS = 600;
		this.ACTIVATION_LENGTH = 40;
		this.Email = $b.email;
		this.LastPatreonEmail = $b.lastPatreonEmail;
		this.IsPatreonSupporter = $b.isPatreonSupporter;
		this.PatreonID = $b.patreonId;
		this.LastPatreonPledgeCents = $b.lastPatreonPledgeCents;
		this.LastTotalCents = $b.lastTotalCents;
		this.MinimumTotalUnits = $b.minimumTotalUnits;
		this.ExternalCents = $b.externalCents;
		this.LastExternalCents = $b.lastExternalCents;
		this.HasSupported = $b.hasSupported;
		this.LastIsAnorak = $b.lastIsAnorak;
		this.RewardType = $b.rewardType;
		this.CustomTier = $b.customTier;
		this.PriorityIssue = $b.priorityIssue;
		/**@type {Date} */
		this.LastActivationTime =
			$b.lastActivationTime instanceof Date
				? $b.lastActivationTime
				: new Date($b.lastActivationTime || 0);
		this.LastPaidPledgeAmount = $b.lastPaidPledgeAmount;
		if ($b.snapshots instanceof List) {
			this.Snapshots = $b.snapshots;
		} else {
			let Snapshots = new List();
			if (Array.isArray($b.snapshots)) {
				for (let snapshot of $b.snapshots) {
					Snapshots.Add(
						snapshot instanceof PatreonSnapshot
							? snapshot
							: new PatreonSnapshot(snapshot)
					);
				}
			}
			this.Snapshots = Snapshots;
		}
	}
	/**
	 * @returns {AccountType}
	 * @readonly
	 * @memberof UserPatreonData
	 */
	get AccountName() {
		if (this.CustomTier != null) return this.CustomTier;
		return this.LastPaidPledgeAmount == 6900 ||
			this.LastPatreonPledgeCents == 6900
			? "Nice."
			: NeosAccount.AccountName(this.CurrentAccountType);
	}
	/**
	 * @returns {AccountType}
	 * @readonly
	 * @memberof UserPatreonData
	 */
	get CurrentAccountType() {
		if (
			new Date(new Date() - this.LastActivationTime).getSeconds() /
				(1000 * 3600 * 24) <=
			40.0
		)
			return UserPatreonData.GetAccountType(this.LastPaidPledgeAmount);
		return AccountType.Normal;
	}
	get CurrentAccountCents() {
		return Math.floor((new Date() - this.LastActivationTime) / 86400000) <= 40
			? this.LastPaidPledgeAmount
			: 0;
	}
	get PledgedAccountType() {
		return UserPatreonData.GetAccountType(this.LastPatreonPledgeCents);
	}
	/**.
	 * //TODO REDO
	 *
	 * @public
	 * @param {number} currentTotalCents
	 * @param currentTotalCents
	 * @param currencyRate
	 * @param findMatchingPledge
	 * @param {Out<boolean>} extendedPlus
	 *
	 * @memberof UserPatreonData
	 */
	UpdatePatreonStatus(
		currentTotalCents,
		currencyRate,
		findMatchingPledge,
		extendedPlus
	) {
		if (
			currentTotalCents < this.MinimumTotalUnits &&
			this.MinimumTotalUnits > 0
		) {
			currentTotalCents = this.MinimumTotalUnits;
			currencyRate = 1.0;
		}
		extendedPlus.Out = false;
		let flag = false;
		if (this.LastTotalUnits === 0 && this.LastTotalCents > 0) {
			this.LastTotalUnits = this.LastTotalCents;
			flag = true;
		}
		let num1 = currentTotalCents - this.LastTotalUnits;
		//TODO Approximate Currency Rate
		let num2 = num1 + (this.ExternalCents - this.LastExternalCents);
		if (num2 <= 0) {
			if (this.LastActivationTime.getFullYear > 2016) return flag;
			num2 = this.LastPaidPledgeAmount;
		}
		if (num2 <= 0) return flag;
		this.LastActivationTime = new Date();
		this.LastPaidPledgeAmount = num2;
		extendedPlus.Out = true;
		this.LastTotalCents += num2;
		this.LastTotalUnits = currentTotalCents;
		this.LastExternalCents = this.ExternalCents;
		this.UpdateMetadata();
		return true;
	}
	UpdateMetadata() {
		this.HasSupported = this.LastTotalCents > 0;
		this.LastIsAnorak = this.LastTotalCents >= 50000;
	}
	/**
	 *
	 * @private
	 * @param {number} cents
	 * @memberof UserPatreonData
	 * @returns {AccountType}
	 */
	GetAccountType(cents) {
		for (var type = AccountType.Anorak; type >= AccountType.Normal; type--) {
			if (cents >= NeosAccount.MinCents(type)) return type;
		}
		return AccountType.Normal;
	}
	get HasPledgesEnoughForPlus() {
		return (
			Math.max(this.LastPatreonPledgeCents, this.LastPaidPledgeAmount) >
			NeosAccount.MinCents(AccountType.BladeRunner)
		);
	}
	get HasPledgedEnoughForWorlds() {
		return (
			Math.max(this.LastPatreonPledgeCents, this.LastPaidPledgeAmount) >= 600
		);
	}
}
module.exports = {
	UserPatreonData,
};
